### PR TITLE
everything

### PR DESCRIPTION
**Pull request by davidchambers**
TODO:

  - [x] Category laws
  - [x] ChainRec laws
  - [x] Comonad laws
  - [x] Contravariant laws
  - [x] Ord laws
  - [x] Profunctor laws
  - [x] Semigroupoid laws
  - [x] Verify dependencies as discussed in fantasyland/fantasy-land#243
  - [x] Revert to ES5 as requested in #2
  - [x] __README.md__

### CODE REVIEW

File: src/ChainRec.js
Comment (davidchambers) [inline src/ChainRec.js]: As with all things ChainRec I'll defer to you on this, @safareli. Should `p`, `n`, and `d` be provided by the caller?

File: src/Comonad.js
Comment (davidchambers) [inline src/Comonad.js]: What is the name of this law?

File: src/Comonad.js
Comment (davidchambers) [inline src/Comonad.js]: What is the name of this law?

File: package.json
Comment (davidchambers) [inline package.json]: We must bump this to `"3.1.x"` once v3.1.0 has been published in order to have access to `Z.contramap`.

File: src/ChainRec.js
Comment (safareli) [reply src/ChainRec.js]: This variables stand for:
- **p**-redicate (to detect if we are done)
- calculate-**n**-extValue (if we are not done, this calculates next value for the iteration)
- calculate-**d**-oneValue (if we are done calculate done value and this will be result of whole iteration)
- **i**-nitialValue (initial value of the iteration)

It's used like this [in FL/test.js](https://github.com/fantasyland/fantasy-land/blob/master/test.js#L58-L64). 

Different monads might use different variation of arguments so I think yes, they should be provided by caller.

File: src/Comonad.js
Comment (safareli) [reply src/Comonad.js]: this laws are same as ones from: [Control/Comonad.purs](https://github.com/purescript/purescript-control/blob/f14399dcd657739745de74cc2cf4ebc3595aadee/src/Control/Comonad.purs#L18-L19)
So i guess it's Left Identity and the other is Right Identity. 
(we need to add names to FL too)

File: src/Comonad.js
Comment (safareli) [reply src/Comonad.js]: Right Identity

File: src/Comonad.js
Comment (davidchambers) [reply src/Comonad.js]: Thank you, Irakli!

Comment (safareli) [reply src/Comonad.js]: Thank you for this PR!

Comment (davidchambers) [reply src/Comonad.js]: fantasyland/fantasy-land#231

File: package.json
Comment (davidchambers) [reply package.json]: :zap:

File: src/Ord.js
Comment (davidchambers) [inline src/Ord.js]: @gabejohnson and @safareli, do these look good to you?

Comment (safareli) [reply src/Ord.js]: lgtm

File: src/Semigroupoid.js
Comment (davidchambers) [inline src/Semigroupoid.js:22]: Please review this file, @gabejohnson.

File: src/Category.js
Comment (davidchambers) [inline src/Category.js:28]: Please review this file, @gabejohnson.

I *think* I ordered the operands to match the spec (going from `_.x(_)` to `x(_, _)` is always confusing). Having done so, it looks as though the law the spec calls left identity is actually right identity, and vice versa.

File: src/ChainRec.js
Comment (davidchambers) [inline src/ChainRec.js:18]: Is it possible to describe this law with an equivalence such as `(a <|> b) <|> c = a <|> (b <|> c)`, @safareli?

**Comment by davidchambers** [comment]
I've written a short readme. Please read it and suggest improvements. :)

File: src/ChainRec.js
Comment (safareli) [reply src/ChainRec.js:18]: Not sure what you mean.

It's already described using equivalence i.e. using ChainRec should be equivalent to using chain in same way

File: src/Category.js
Comment (gabejohnson) [reply src/Category.js:28]: `a.compose(C.id())` is equivalent to `a` (right identity)
`C.id().compose(a)` is equivalent to `a` (left identity)

These look right to me. The identity element `C.id()` is on the right for "right identity" and on the left for "left identity".

The issue is `compose`. The arguments are swapped in the dispatch function (as they should be). See https://github.com/fantasyland/fantasy-land/pull/249

File: src/Semigroupoid.js
Comment (gabejohnson) [reply src/Semigroupoid.js:22]: LGTM

File: src/ChainRec.js
Comment (davidchambers) [reply src/ChainRec.js:18]: I was wondering whether

```javascript
M.chainRec((next, done, v) => p(v) ? d(v).map(done) : n(v).map(next), i)
= (function step(v) { return p(v) ? d(v) : n(v).chain(step); }(i))
```

can be expressed in the Haskellesque syntax we're using to describe the other laws. I don't think this is possible.

File: src/Category.js
Comment (davidchambers) [reply src/Category.js:28]: Thanks for checking, Gabe. It was this that looked wrong to me:

```console
$ ack --before-context 1 --no-filename -- leftIdentity src | awk 'NR % 3 == 1'
    //  a `compose` id C = a
    //  extend extract w = w
    //  pure x >>= f = f x
    //  empty <> m = m
    //  zero <|> x = x
```

In the cases of `<>` and `<|>` at least, the identity element is the left operand.

File: README.md
Comment (Avaq) [inline README.md]: `!`?

File: src/ChainRec.js
Comment (safareli) [reply src/ChainRec.js:18]: We can covert it to Haskellesque syntax, but it still will be verbose

**Comment by gabejohnson** [comment]
Very helpful David! Thank you.

File: README.md
Comment (davidchambers) [reply README.md]: :zap:

File: src/ChainRec.js
Comment (davidchambers) [reply src/ChainRec.js:18]: Let's just skip this one. :)

**Comment by davidchambers** [comment]
@dakom, would you mind reviewing the updated readme? I hope it's clearer now that it shows how to test a property of a custom type rather than a property of a built-in type.

**Review by dakom — COMMENTED** [review]
The example and language changes are great, thanks for doing that!

I'll need to try implementing it by following the directions before I can offer more detailed feedback, but on one foot I do have two suggestions to consider:

> Provide the fixed parameters to laws.Group:

It might help to have a note there about how you knew to use `laws.Group` and how to find which functions are exported. This isn't such a hurdle since one already needs to know which category they intend to test against, but I think it could use a bit more explicit info. I don't feel strongly about this one though - just a minor point.

> Define an "arbitrary" for the type

This is the most confusing to me - though it probably just requires more familiarity with jsverify which is outside the scope of Sanctuary.

Still - anecdotally, I find it difficult to see how you get from the jsverify explanation:

"g should be a right inverse of f, but doesn't need to be complete inverse. i.e. i.e. f doesn't need to be invertible, only surjective." (on the linked page) 

to:
`const SumArb = jsc.number.smap(Sum, sum => sum.value, Z.toString);`

Establishing that connection in a bit more detail would help me know how to implement it for other types, I think... though of course some of that responsibility falls on the jsverify side ;)

**Comment by davidchambers** [comment]
@Avaq, @gabejohnson, or @safareli, would one of you be so kind as to give this pull request one more look before I merge it?

**Comment by davidchambers** [comment]
I'm sorry if you received lots of notifications as I was wrestling with Travis. All is well now. :)

**Review by Avaq — APPROVED** [review]

**Pull request merged**
